pkg/k8s/watchers/subscriber/node.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package subscriber
+
+import (
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+)
+
+// NodeHandler is implemented by event handlers responding to K8s Node events.
+type NodeHandler interface {
+	OnAddNode(*slim_corev1.Node) error
+	OnUpdateNode(oldObj, newObj *slim_corev1.Node) error
+	OnDeleteNode(*slim_corev1.Node) error
+}
+
+// NewNode creates a new subscriber list for NodeHandlers.
+func NewNode() *NodeList {
+	return &NodeList{}
+}
+
+// Register registers NodeHandler as a subscriber for reacting to Node objects
+// into the list.
+func (l *NodeList) Register(s NodeHandler) {
+	l.Lock()
+	l.subs = append(l.subs, s)
+	l.Unlock()
+}
+
+// NotifyAdd notifies all the subscribers of an add event to a service.
+func (l *NodeList) NotifyAdd(node *slim_corev1.Node) []error {
+	l.RLock()
+	defer l.RUnlock()
+	errs := make([]error, 0, len(l.subs))
+	for _, s := range l.subs {
+		if err := s.OnAddNode(node); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+// NotifyUpdate notifies all the subscribers of an update event to a service.
+func (l *NodeList) NotifyUpdate(oldNode, newNode *slim_corev1.Node) []error {
+	l.RLock()
+	defer l.RUnlock()
+	errs := make([]error, 0, len(l.subs))
+	for _, s := range l.subs {
+		if err := s.OnUpdateNode(oldNode, newNode); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+// NotifyDelete notifies all the subscribers of an update event to a service.
+func (l *NodeList) NotifyDelete(node *slim_corev1.Node) []error {
+	l.RLock()
+	defer l.RUnlock()
+	errs := make([]error, 0, len(l.subs))
+	for _, s := range l.subs {
+		if err := s.OnDeleteNode(node); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+// NodeList holds the NodeHandler subscribers that are notified when reacting
+// to K8s Node resource / object changes in the K8s watchers.
+type NodeList struct {
```

Comment:
@christarazi one other naming suggestion, to keep semantics related to pub/sub how do you feel about calling this a "chain" instead of "list". 

Im loosely borrowing the terms used in notification chain semantics here: https://0xax.gitbooks.io/linux-insides/content/Concepts/linux-cpu-4.html as a model.

Reply:
I'm fine with this renaming :+1:

Reply:
@christarazi one other note here:

The "NodeChain" could itself implement "subscriber.Node" interface - this would be similar to how golang's ServeMux implements http.Handler. The serve mux itself is a handler, which will register and call other handlers. see: https://golang.org/src/net/http/server.go?s=76212:76272#L2429

In the same vein the "NodeChain" itself can be a subscriber.Node - allowing the method set to stay the same across k8sWatcher usage and implementations.